### PR TITLE
feat: add demo page-level language selection

### PR DESCRIPTION
**Pull request by dlockhart**
Builds on Joseph's "language toggle" but moves it up to be a page-level dropdown.

I've added logic here that sets the document's direction based on the language chosen, which removes the need for a separate "RTL" switch. That switch was pretty buggy anyway. This did require a modification to the `RtlMixin` so that it listens for changes to the document language and re-calculates based on that.

English:
Screen Shot 2022-11-25 at 2 11 36 PM

Arabic:
Screen Shot 2022-11-25 at 2 11 47 PM

### CODE REVIEW

File: components/demo/demo-page.js [hunk 1]
```diff
@@ -86,6 +109,13 @@ class DemoPage extends LitElement {
 			this._noScroll = true;
 		}
 	}
+
+	_handleLanguageChange(e) {
+		const newLanguageCode = e.target[e.target.selectedIndex].value;
+		document.documentElement.dir = newLanguageCode === 'ar' ? 'rtl' : 'ltr';
```
Comment (dlockhart) [inline components/demo/demo-page.js:115]: If we like this approach, I'll make a separate change to the Intl library so that when you set the language to Arabic, it automatically sets the document's direction.

File: components/demo/demo-page.js [hunk 2]
```diff
@@ -67,9 +80,19 @@ class DemoPage extends LitElement {
 		const classes = {
 			'no-scroll': this._noScroll
 		};
+		const selectedLanguageCode = getDocumentLocaleSettings().language;
+		const languageOptions = supportedBaseLocales.map((code) => {
```
Comment (dlockhart) [inline components/demo/demo-page.js:84]: This outputs the not super friendly language code... I could also add something to the Intl library that exposes a friendly list of locale names. If we like this approach, I can do that.

File: mixins/rtl-mixin.js [hunk 3]
```diff
@@ -13,9 +14,25 @@ export const RtlMixin = dedupeMixin(superclass => class extends superclass {
...
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		this._localeSettings.removeChangeListener(this._handleLanguageChange);
+	}
+
+	_handleLanguageChange() {
 		const dir = document.documentElement.getAttribute('dir');
 		// avoid reflecting "ltr" for better performance
-		if (dir && dir !== 'ltr') {
+		if (dir && (dir !== 'ltr' || this.dir === 'rtl')) {
```
Comment (dlockhart) [inline mixins/rtl-mixin.js:35]: This logic needed slight tweaking to handle a change after the initial value was set. If things go from `rtl` -> `ltr`, we DO need to reflect that change. I could also set it back to `undefined` -- what do folks think?

Comment (dbatiste) [reply mixins/rtl-mixin.js:35]: Is it valid to assign `this.dir = undefined`? When I log `dir` for elements without an explicit `dir` assignment, the browser returns an empty string, though [Mozilla says](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dir) the property's values can be `ltr`, `rtl`, and `auto`.

File: mixins/rtl-mixin.js [hunk 4]
```diff
@@ -13,9 +14,25 @@ export const RtlMixin = dedupeMixin(superclass => class extends superclass {
...
 	constructor() {
 		super();
+		this._localeSettings = getDocumentLocaleSettings();
+		this._handleLanguageChange = this._handleLanguageChange.bind(this);
+		this._handleLanguageChange();
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+		this._localeSettings.addChangeListener(this._handleLanguageChange);
```
Comment (dbatiste) [inline mixins/rtl-mixin.js:24]: Would these listeners only get invoked if the language changes after the page loads? i.e. in a normal page where the language doesn't change, would `_handleLanguageChange` only get called once, from the `constructor`?

I'm not sure how concerned we should be about the number of handlers wired up. I can imagine pages easily wiring up hundreds due to the number of components, perhaps even more than a thousand. If the change handler is only needed for pages such as the demo page where the language can change, we could do this wire-up conditionally based on some global flag that the demo page sets. Ideally we can just wire it up and not worry about it though.

**Review by dbatiste — COMMENTED** [review]
I like it! Just a couple comments/questions.

File: mixins/rtl-mixin.js (hunk 4 above)
Comment (dlockhart) [reply mixins/rtl-mixin.js:24]: Correct, these only get called if the language changes after the page loads. That currently only happens in the demo page (and maybe a few of our tests), although it's not impossible to foresee a scenario where we may allow the user to switch languages without reloading the whole app. Pretty edge case-y though.

I'm not especially worried about the add/remove of the listeners... it's just [appending to an array and then splicing it](https://github.com/BrightspaceUI/intl/blob/main/lib/common.js#L129. Agreed though that we could do some kind of global flag if it becomes an issue.

What I'm hoping though is that our dependence on `RtlMixin` will actually start to drop off now that we're in the clear to start using CSS logical properties. I'm guessing 95% of our uses are related to that.

File: mixins/rtl-mixin.js (hunk 3 above)
Comment (dlockhart) [reply mixins/rtl-mixin.js:35]: Hmm, interesting. I just checked, and even before these changes, any component using `RtlMixin` was having its `dir` set to `undefined` by default. I'm guessing Lit's doing that since we're redefining `dir` as a property and aren't initializing it in the constructor.

If we're worried about it being `undefined`, we could _not_ define it as a Lit property (since it's already a native one) and just manually call `requestUpdate` when we set it.

Comment (dbatiste) [reply mixins/rtl-mixin.js:35]: Cool - I'm happy either way as long as the browser doesn't 🤮 .

File: components/demo/demo-page.js (hunk 1 above)
Comment (dlockhart) [reply components/demo/demo-page.js:115]: Actually on second thought, the Intl library doesn't set the `lang` on the `<html>` element so I don't think this makes sense. It just listens for mutations to those attributes and adapts accordingly.

File: components/demo/demo-page.js [hunk 5]
```diff
@@ -3,8 +3,11 @@ import './code-view.js';
 import '../colors/colors.js';
 import '../typography/typography.js';
 import { css, html, LitElement } from 'lit';
+import { getDocumentLocaleSettings, supportedBaseLocales } from '@brightspace-ui/intl/lib/common.js';
```
Comment (dbatiste) [inline components/demo/demo-page.js:6]: Would you want to use the base locales or the culture-specific ones?

Comment (dlockhart) [reply components/demo/demo-page.js:6]: I can do either, but I figured the base locales accomplished the task. Is there any advantage to the culture-specific ones?

Comment (dbatiste) [reply components/demo/demo-page.js:6]: I think either should be fine. I guess the only thing would be if someone wants to check the verbiage for a culture specific locale.

Comment (dlockhart) [reply components/demo/demo-page.js:6]: Yeah. Ok I'll use the larger set -- need to update my PR in Intl then.

Comment (dlockhart) [reply components/demo/demo-page.js]: Ok -- updated with the culture-specific ones. Screenshots in description updated as well.

**Review by dbatiste — APPROVED** [review]

File: components/demo/demo-page.js [hunk 6]
```diff
@@ -86,6 +112,13 @@ class DemoPage extends LitElement {
 			this._noScroll = true;
 		}
 	}
+
+	_handleLanguageChange(e) {
+		const newLanguageCode = e.target[e.target.selectedIndex].value;
+		document.documentElement.dir = newLanguageCode === 'ar-sa' ? 'rtl' : 'ltr';
```
Comment (eKoopmans) [inline components/demo/demo-page.js:118]: So now to test RTL visually, we'll need to explicitly select `ar-sa`. Sounds good, might be worth calling out on our Slack or something though 😄.

**Review by eKoopmans — APPROVED** [review]
